File: TTT.py

grid = [[" " for _ in range(3)] for _ in range(3)]

def draw_grid():
    print()
    for row in grid:
        print(f"| {row[0]} | {row[1]} | {row[2]} |")
    print()

def menu():
    print("_" * 25)
    print("Block placements are:-")
    print("|1|2|3|")
    print("|4|5|6|")
    print("|7|8|9|")
    print("Use these numbers to place your mark!")
    print("_" * 25)

def location(i):
    """Map input 1-9 to row, col indexes"""
    mapping = {
        1: (0, 0), 2: (0, 1), 3: (0, 2),
        4: (1, 0), 5: (1, 1), 6: (1, 2),
        7: (2, 0), 8: (2, 1), 9: (2, 2)
    }
    return mapping.get(i, None)

def Pinput(player):
    """Get valid input for player"""
    while True:
        try:
            l = int(input(f"Player {player} input (1-9): "))
            pos = location(l)
            if not pos:
                print("❌ Invalid choice! Pick between 1-9.")
                continue
            if grid[pos[0]][pos[1]] != " ":
                print("❌ Cell already taken, choose another!")
                continue
            return pos
        except ValueError:
            print("❌ Invalid input! Please enter a number.")

def won(g):
    """Check if someone won"""
    lines = (
        # Rows
        [(0, 0), (0, 1), (0, 2)],
        [(1, 0), (1, 1), (1, 2)],
        [(2, 0), (2, 1), (2, 2)],
        # Columns
        [(0, 0), (1, 0), (2, 0)],
        [(0, 1), (1, 1), (2, 1)],
        [(0, 2), (1, 2), (2, 2)],
        # Diagonals
        [(0, 0), (1, 1), (2, 2)],
        [(0, 2), (1, 1), (2, 0)],
    )
    for line in lines:
        a, b, c = line
        if g[a[0]][a[1]] == g[b[0]][b[1]] == g[c[0]][c[1]] != " ":
            return True, g[a[0]][a[1]]
    return False, None

def is_full(g):
    return all(cell != " " for row in g for cell in row)

def main():
    menu()
    draw_grid()

    current_player = 1
    symbol = {1: "X", 2: "O"}

    while True:
        pos = Pinput(current_player)
        grid[pos[0]][pos[1]] = symbol[current_player]
        draw_grid()

        w, winner = won(grid)
        if w:
            print(f"🎉 Winner is Player {current_player} ({winner})!")
            break
        if is_full(grid):
            print("🤝 It's a draw!")
            break

        # Switch player
        current_player = 2 if current_player == 1 else 1

if __name__ == "__main__":
    main()

File: ttt.cpp
#include <iostream>
#include <limits>

using namespace std;

char grid[3][3] = {  {'.', '.', '.'}
                    ,{'.', '.', '.'}
                    ,{'.', '.', '.'}};
int rows = sizeof(grid)/sizeof(grid[0]);
int columns = sizeof(grid[0])/sizeof(grid[0][0]);


void drawGrid(){
    for(int i = 0; i < rows; i++){
        for(int j = 0; j < columns; j++){
            cout<<"| " << grid[i][j];
        }
        cout<<"|"<<endl;
    }
}


void menu(){
    cout << "\nSelect a number from this refrence grid to give input:- \n";
    cout << "1|2|3|\n";
    cout << "4|5|6|\n";
    cout << "7|8|9|\n";
}

void updateGrid(int choice, char symbol) {
    if (choice < 1 || choice > 9) {
        cout << "Invalid move!" << endl;
        return;
    }

    int row = (choice - 1) / 3; // map 1-9 → row
    int col = (choice - 1) % 3; // map 1-9 → col

    if (grid[row][col] == '.') { // check if empty
        grid[row][col] = symbol;
    } else {
        cout << "Block already taken!" << endl;
    }
    drawGrid();
    cout << endl;
}

int Input(){
    int p=0;
    cin >> p;

    return p;
}

bool isEmpty() {
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < columns; j++) {
            if (grid[i][j] == '.') {
                return true; // found at least one empty space
            }
        }
    }
    return false; // no empty spaces left
}

bool CheckWin() {
    // Check rows
    for (int i = 0; i < rows; i++) {
        if (grid[i][0] != '.' && grid[i][0] == grid[i][1] && grid[i][1] == grid[i][2]) {
            cout << "Player " << (grid[i][0] == 'X' ? "1 (X)" : "2 (O)") << " wins!\n";
            return true;
        }
    }

    // Check columns
    for (int j = 0; j < columns; j++) {
        if (grid[0][j] != '.' && grid[0][j] == grid[1][j] && grid[1][j] == grid[2][j]) {
            cout << "Player " << (grid[0][j] == 'X' ? "1 (X)" : "2 (O)") << " wins!\n";
            return true;
        }
    }

    // Check main diagonal
    if (grid[0][0] != '.' && grid[0][0] == grid[1][1] && grid[1][1] == grid[2][2]) {
        cout << "Player " << (grid[0][0] == 'X' ? "1 (X)" : "2 (O)") << " wins!\n";
        return true;
    }

    // Check anti-diagonal
    if (grid[0][2] != '.' && grid[0][2] == grid[1][1] && grid[1][1] == grid[2][0]) {
        cout << "Player " << (grid[0][2] == 'X' ? "1 (X)" : "2 (O)") << " wins!\n";
        return true;
    }

    return false; // no winner yet
}



int main() {
    cout << "///////////////////////////////////////////////////////////////////////\n";
    menu();
    do {
        cout << "///////////////////////////////////////////////////////////////////////\n";
        cout << "Player 1 (X) turn: ";
        int player1 = Input();
        updateGrid(player1, 'X');
        if (CheckWin()) break; // check after player 1 move

        if (!isEmpty()) break; // stop if board is full

        cout << "///////////////////////////////////////////////////////////////////////\n";
        cout << "Player 2 (O) turn: ";
        int player2 = Input();
        updateGrid(player2, 'O');
        if (CheckWin()) break; // check after player 2 move

    } while (isEmpty());

    if (!CheckWin()) {
        cout << "It's a draw!\n";
    }

    // Pause before exit
    cout << "\nPress Enter to exit...";
    cin.ignore(numeric_limits<streamsize>::max(), '\n'); // clear leftover input
    cin.get(); // wait for Enter

    return 0;
}
